pkg/pool-gyro/contracts/GyroECLPPool.sol

```diff
@@ -0,0 +1,229 @@
+ * parameterized by the pricing range [α,β], the inclination angle `phi` and stretching parameter `lambda`. For more
+ * information, please refer to https://docs.gyro.finance/gyroscope-protocol/concentrated-liquidity-pools/e-clps.
+ */
+contract GyroECLPPool is IGyroECLPPool, BalancerPoolToken {
+    using FixedPoint for uint256;
+    using SafeCast for *;
+
+    /// @dev Parameters of the ECLP pool
+    int256 internal immutable _paramsAlpha;
+    int256 internal immutable _paramsBeta;
+    int256 internal immutable _paramsC;
+    int256 internal immutable _paramsS;
+    int256 internal immutable _paramsLambda;
+
+    /**
+     * @dev Derived Parameters of the E-CLP pool, calculated off-chain based on the parameters above. 38 decimals
+     * precision.
+     */
+    int256 internal immutable _tauAlphaX;
+    int256 internal immutable _tauAlphaY;
+    int256 internal immutable _tauBetaX;
+    int256 internal immutable _tauBetaY;
+    int256 internal immutable _u;
+    int256 internal immutable _v;
+    int256 internal immutable _w;
+    int256 internal immutable _z;
+    int256 internal immutable _dSq;
+
+    bytes32 private constant _POOL_TYPE = "ECLP";
+
+    constructor(GyroECLPPoolParams memory params, IVault vault) BalancerPoolToken(vault, params.name, params.symbol) {
+        GyroECLPMath.validateParams(params.eclpParams);
+        emit ECLPParamsValidated(true);
+
+        GyroECLPMath.validateDerivedParamsLimits(params.eclpParams, params.derivedEclpParams);
+        emit ECLPDerivedParamsValidated(true);
+
+        (_paramsAlpha, _paramsBeta, _paramsC, _paramsS, _paramsLambda) = (
+            params.eclpParams.alpha,
+            params.eclpParams.beta,
+            params.eclpParams.c,
+            params.eclpParams.s,
+            params.eclpParams.lambda
+        );
+
+        (_tauAlphaX, _tauAlphaY, _tauBetaX, _tauBetaY, _u, _v, _w, _z, _dSq) = (
+            params.derivedEclpParams.tauAlpha.x,
+            params.derivedEclpParams.tauAlpha.y,
+            params.derivedEclpParams.tauBeta.x,
+            params.derivedEclpParams.tauBeta.y,
+            params.derivedEclpParams.u,
+            params.derivedEclpParams.v,
+            params.derivedEclpParams.w,
+            params.derivedEclpParams.z,
+            params.derivedEclpParams.dSq
+        );
+    }
+
+    /// @inheritdoc IBasePool
+    function computeInvariant(
+        uint256[] memory balancesLiveScaled18,
+        Rounding rounding
+    ) external view returns (uint256) {
+        (EclpParams memory eclpParams, DerivedEclpParams memory derivedECLPParams) = _reconstructECLPParams();
+
+        (int256 currentInvariant, int256 invErr) = GyroECLPMath.calculateInvariantWithError(
+            balancesLiveScaled18,
+            eclpParams,
+            derivedECLPParams
+        );
+
+        if (rounding == Rounding.ROUND_DOWN) {
+            return currentInvariant.toUint256();
```

Comment:
I subtracted the error when rounding down, since the tests are passing, but asked them anyway. If they give a reason not to do it, I revert this refactor.